File: two_sum/two_sum.js
/*
    @param {Integer[]} nums
    @param {Integer} target
    @return {Integer[]}
    
    plan: 
    Populate a object where the keys are numbers and the values are the indicies of their compliments
    to the target.
*/


function twoSum(nums, target){
    let complimentsObject = {}
    for (let i = 0; i < nums.length; i++){
        if (complimentsObject[nums[i]] !== undefined){
            return [complimentsObject[nums[i]], i];
        }
        complimentsObject[target - nums[i]] = i;
    }
    return false;
}
/*
    test case: 
    Given nums = [2, 7, 11, 15], target = 9,
    Because nums[0] + nums[1] = 2 + 7 = 9,
    return [0, 1].
*/

const nums = [2, 7, 11, 15]
const target = 9

console.log(twoSum(nums, target)); //expect [ 0 , 1 ]

File: two_sum/two_sum.rb
=begin
    @param {Integer[]} nums
    @param {Integer} target
    @return {Integer[]}
    
    plan: 
    Populate a hash where the keys are numbers and the values are the indicies of their compliments
    to the target.
=end


def two_sum(nums, target)
    compliments_hash = Hash.new(false)
    nums.each_with_index do |num, index| 
        if (compliments_hash[target - num])
            return [compliments_hash[target - num], index]
        end
        compliments_hash[num] = index;
    end
    false
end

=begin
    test case: 
    Given nums = [2, 7, 11, 15], target = 9,
    Because nums[0] + nums[1] = 2 + 7 = 9,
    return [0, 1].
=end

nums = [2, 7, 11, 15]
target = 9

puts two_sum(nums, target) == [0, 1]